### PR TITLE
Better peer selection

### PR DESCRIPTION
**Pull request by evanxg852000**
- Fixed peer selection logic.
- Guard against network partition

https://github.com/quickwit-oss/scuttlebutt/issues/11

### CODE REVIEW

File: scuttlebutt/src/server.rs [hunk 1]
```diff
@@ -172,30 +173,113 @@ impl UdpServer {
...
         const EMPTY_STRING: String = String::new();
         let mut rand_nodes = [EMPTY_STRING; GOSSIP_COUNT];
 
-        // Select up to [`GOSSIP_COUNT`] node IDs at random.
-        let nodes = scuttlebutt_guard.cluster_state.nodes();
-        let count = nodes
-            .filter(|node_id| node_id != self_node_id)
+        // Select `GOSSIP_COUNT` number of live nodes.
+        // On startup, select from cluster nodes since we don't know any live node yet.
+        let count = if live_nodes_count == 0 {
```
Comment (guilload) [inline scuttlebutt/src/server.rs:185]: Would it be simpler to consider the seed nodes live on startup?

File: scuttlebutt/src/failure_detector.rs [hunk 2]
```diff
@@ -69,20 +72,27 @@ impl FailureDetector {
...
             } else {
                 self.live_nodes.insert(node_id.to_string());
+                self.dead_nodes.remove(node_id);
             }
         }
     }
 
-    /// Returns a list of living nodes.
+    /// Returns a list of live nodes.
     pub fn live_nodes(&self) -> impl Iterator<Item = &str> {
```
Comment (guilload) [inline scuttlebutt/src/failure_detector.rs:87]: We almost always end up collecting into a set. Should we return a `HashSet<NodeId>` directly for `live_nodes`, `dead_nodes`, `seed_nodes`?

File: scuttlebutt/src/server.rs [hunk 3]
```diff
@@ -172,30 +173,113 @@ impl UdpServer {
...
         // Update nodes liveliness
         let mut scuttlebutt_guard = self.scuttlebutt.lock().await;
         scuttlebutt_guard.update_nodes_liveliness();
     }
 
+    /// Selects a dead node to gossip with, with some probability.
+    fn select_dead_node_to_gossip_with(
+        &self,
+        rng: &mut SmallRng,
+        dead_nodes: &HashSet<String>,
```
Comment (guilload) [inline scuttlebutt/src/server.rs:251]: Can we use `NodeId` whenever possible?

File: scuttlebutt/src/server.rs [hunk 4]
```diff
@@ -172,30 +173,113 @@ impl UdpServer {
...
+            dead_nodes_count,
+        );
 
         // Drop lock to prevent deadlock in [`UdpSocket::gossip`].
         drop(scuttlebutt_guard);
+        let mut has_gossiped_with_a_seed_node = false;
         for node in &rand_nodes[..count] {
+            has_gossiped_with_a_seed_node =
+                has_gossiped_with_a_seed_node || seed_nodes.contains(node);
             let _ = self.gossip(node).await;
```
Comment (guilload) [inline scuttlebutt/src/server.rs:230]: Can we log potential errors?

**Review by guilload — COMMENTED** [review]
Since the complexity of the peer selection logic has increased, I'd isolate that logic from `gossip_multiple` so we can test it more easily.

File: scuttlebutt/src/server.rs (hunk 1 above)
Comment (evanxg852000) [reply scuttlebutt/src/server.rs:185]: The `cluster_nodes` will most likely contain seed nodes so I felt it's better. Besides this is just a case for the startup, in normal operation, we want live nodes.

File: scuttlebutt-test/Cargo.toml [hunk 5]
```diff
@@ -21,4 +22,4 @@ tracing-subscriber = "0.3"
 assert_cmd = "2"
 predicates = "2"
 reqwest = { version = "0.11", default-features=false, features = ["blocking", "json"] }
-
+serial_test = "0.5.1"
```
Comment (guilload) [inline scuttlebutt-test/Cargo.toml:25]: Let's implement tests that can be run in parallel using random ports. Two tests can run parallel using random ports in different ranges.

File: scuttlebutt/src/server.rs [hunk 6]
```diff
@@ -172,23 +174,42 @@ impl UdpServer {
...
-
-        // Select up to [`GOSSIP_COUNT`] node IDs at random.
-        let nodes = scuttlebutt_guard.cluster_state.nodes();
-        let count = nodes
-            .filter(|node_id| node_id != self_node_id)
-            .map(ToString::to_string)
-            .choose_multiple_fill(rng, &mut rand_nodes);
+        let cluster_state = scuttlebutt_guard.cluster_state();
+
+        let all_nodes = cluster_state
```
Comment (guilload) [inline scuttlebutt/src/server.rs:181]:
```suggestion
        let peer_nodes = cluster_state
```

File: scuttlebutt/src/state.rs [hunk 7]
```diff
@@ -90,12 +90,14 @@ impl NodeState {
 
 #[derive(Default, Serialize, Deserialize, Clone, Debug)]
 pub struct ClusterState {
+    seed_nodes: Vec<String>,
```
Comment (guilload) [inline scuttlebutt/src/state.rs:93]:
```suggestion
    seed_nodes: HashSet<NodeId>,
```

**Review by guilload — APPROVED** [review]

File: scuttlebutt-test/Cargo.toml (hunk 5 above)
Comment (fulmicoton) [reply scuttlebutt-test/Cargo.toml:25]: Not sure if this works here and universal this trick is, but when listenign to port 0, linux picks a random UDP port that is available.

**Pull request merged**